sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -626,7 +626,7 @@ object Decimal {
       }
     } catch {
       case _: NumberFormatException =>
-        throw QueryExecutionErrors.invalidInputSyntaxForNumericError(str, errorContext)
+        throw QueryExecutionErrors.invalidInputSyntaxForNumericError(DecimalType, str, errorContext)
```

Comment:
Where do we call `fromStringANSI`? There should be a concrete `DecimalType` available

Reply:
I don't think we can get that concrete Decimal type because the cast fails: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala#L629

Reply:
We can get it from the caller:
https://github.com/apache/spark/pull/36244/files#diff-393a0ad195d66fb094f2dca3669c1317b7f57a6637f6e95241b57bc6550ce49aR819

Reply:
I didn't do it initially, because I think `fromStringANSI` and `changePrecision` are two separate processes and has its corresponding exceptions. The former has an exception to cast from a string to a BigDecimal, and at this step, the `target` DecimalType doesn't appear. While the latter has exception to change precision, where `target` starts to take place.

But maybe even though these are two steps, they are always called together serially. Thus, having the `target` information in the first stage seems fine. It also saves a lot of trouble in `AbstractDataType` etc. Will update the PR.